graphql/src/runner.rs

```diff
@@ -140,15 +135,37 @@ where
                     max_first: max_first.unwrap_or(*GRAPHQL_MAX_FIRST),
                     load_manager: self.load_manager.clone(),
                 },
-            ) {
-                Err(errs) => errors.extend(errs),
-                Ok(mut vals) => values.append(&mut vals),
+            )
+        };
+        let by_block_constraint = query.block_constraint()?;
+
+        // We want to optimize for the common case of a single block constraint,
+        // where we can avoid cloning the result.
+        match by_block_constraint.len() {
```

Comment:
A simpler way to special-case the first item is with this pattern...

```rust
let mut by_block_constraint = query.block_constraint()?.into_iter();

let mut result = if let Some((bc, selection_set)) = if by_block_constraint.next() {
     // Do the first one
} else {
  QueryResult::empty()
}

for (bc, selection_set) in by_block_constraint {
  // remainder
}

Ok(Arc::new(result))
```

Also right now the code calls `query.block_constraint()?` again in the `_ => {` arm seemingly unnecessarily.

Reply:
I wasn't happy with how this turned out, that seems much better, I'll try it.